Support axis_index_groups in allreduce collectives

jekbradbury opened the pull request:
This lets users write collective operations that reduce over groups of axis indices within a pmapped axis rather than over the entire axis.
Example:

```python
def sum_with_groups(x):
  return psum(x, axis_name="i", axis_index_groups=[[0, 1], [2, 3]])
x = jnp.array([1., 2., 3., 4.])
y = pmap(sum_with_groups, axis_name="i")(x)  # [3., 3., 7., 7.]
```

Original motivation:
Even in the most composable and multihost-friendly nested-pmap world that we can come up with, I think we'll still want to be able to express collectives that operate over subsets of the pmap axis. In particular I think it's more natural, for example, for only the batchnorm layer to have to care about its distributed group size (whether to reduce over individual devices, neighboring pairs of devices, whole hosts, etc.), and keep the main pmap over the whole device axis, than for the entire model/training loop to need to know that (and use a pair of pmaps with separate axes).

Edit: this now operates at the level of JAX axis indices, so it works correctly with nested pmaps (they lower to XLA replica groups together).

Edit again, quoting from https://github.com/google/jax/pull/2382#issuecomment-625909818:

> I think the stronger version is to point to things we can't express even with nested pmaps (without requiring a sequence of separate pmap calls and some transposition). For example, with this proposal in a single pmap one can write axis_index_groups=[[0, 2], [1, 3]] and in another operation write axis_index_groups=[[0, 1], [2, 3]], right?

jekbradbury committed:
support replica groups in allreduce collectives

jekbradbury committed:
add test and fix jaxpr in docs

skye commented on tests/pmap_test.py:
```diff
@@ -352,6 +352,25 @@ def sum_and_broadcast(x, axis):
+    if replicas % 2 != 0:
+      raise SkipTest
+    replica_groups = onp.arange(replicas).reshape(2, replicas // 2).tolist()
+    f = pmap(lambda x: x - lax.psum(x, 'i', replica_groups), axis_name='i')
```
Do replica_groups work with nested pmaps?

jekbradbury replied:
I'd say no, not really (I mean, it'll work but it'll use the computation-global replica IDs, breaking the nested pmap abstraction). I guess we can either support per-axis replica groups (not sure how hard that would be, but I don't want to underestimate things that touch nested pmap) or we could throw an error if you use replica_groups in a nested pmap?

mattjj reviewed (changes requested):
Can we see a design doc for this? I had some ideas for how to do this (in a conversation with Skye) a few weeks ago.

(If this is just meant to be a short-term hack for a specific use case, we'd still like to do some things differently, like not include it in docstrings, and maybe introduce a new separate primitive instead of changing the existing ones.)

skye commented:
Matt can you (briefly) share your ideas here? I don't remember the specifics of our convo and am very curious :) Plus it might help James with a design doc. No worries if you're busy and wanna save it for a future design discussion.

jekbradbury committed:
switch from XLA replica IDs to JAX axis indices

jekbradbury commented:
Updates since the earlier version:

- I spent some time trying to use nested pmap for the original motivating use case here (distributed batchnorm with reduction scope between "one device" and "global"), since that's the existing mechanism for expressing nested levels of collective computation (and also the existing way to use XLA replica groups). This is possible (I trained ResNet with it) and will be free of excess data movement after #2893, but it's not a particularly ergonomic API for distributed batchnorm.
In particular, using nested pmap means that enabling distributed batchnorm is a training-loop-level change (rather than a change that can be localized to the batchnorm layer); it means that many arrays encountered during the training loop have two leading axes; and it means that _all_ computations that take these doubly-sharded `ShardedDeviceArray`s as input, even ones that have no or only global collectives, need to be wrapped in two pmaps (modulo heroic and possibly confusing things we can do with replication).

- I learned more about the semantics of XLA replica groups, in particular that the order of replicas within a group (but not the order of groups) is currently load-bearing, and can make a big performance difference even on commutative allreduces. (Arguably, canonicalizing the within-group order for optimal physical locality is a missing XLA feature.) As long as this is the case, it's important to give some way for end users to control the within-group order. While the relationship between JAX axis indices and physical device locations is somewhat complex (axis indices from one or more pmap axes are multiplexed into XLA replicas, then replicas are assigned to devices according to the standard JAX multihost device assignment) it is at least stable and (in the absence of nested pmap) as simple as we can get without directly surfacing the physical topology.

- JAXers (including two researchers) patched the original version of this change and found it useful.

- I switched this functionality to operate at the level of JAX axis indices rather than punching a hole through to XLA replica IDs (and renamed the kwarg from `replica_groups` to `axis_index_groups`). This means it now composes with nested pmap.

Reasons for the nested list API over other possible choices:
- it expresses an order within each group, which turned out to be important; using e.g. `[0, 0, 1, 1]` for two groups of two replicas doesn't
- it conveys a notion of "groups" in a way that e.g. a 2D numpy array doesn't (are the groups in the first or second axis?)
- it's visually similar to the XLA replica group specification some users are familiar with from HLO text

jekbradbury committed:
fix psum transpose rule

skye commented on tests/pmap_test.py:
```diff
@@ -384,6 +384,50 @@ def sum_and_broadcast(x, axis):
+    axis_index_groups = onp.arange(replicas // 2).reshape(
+      2, replicas // 4).tolist()
+    f = lambda x: x - lax.psum(x, 'i', axis_index_groups=axis_index_groups)
+    f = pmap(pmap(f, 'i'), 'j')
```
Can you also test `f = pmap(pmap(f, 'j'), 'i')` ? (reverse the `i` and `j` axes) I'm imagining just copy/pasting from here down and changing the axes accordingly

jekbradbury replied:
Done.

skye commented on tests/pmap_test.py:
```diff
@@ -384,6 +384,50 @@ def sum_and_broadcast(x, axis):
+
+    ans = f(x)
+    self.assertAllClose(ans, expected, check_dtypes=False)
+
```
Also if you wanna be super future proof... add test where the pmaps are "perfectly nested" :O (since once we don't use replication for perfectly nested pmaps we'll lose test coverage of that logic, which I just reviewed and is quite tricky!)

Or just wait until we write a fuzzer ;)

jekbradbury replied:
Done (assuming you meant "imperfectly nested", since the others are perfectly nested).

skye replied:
I did mean imperfectly, oops.

skye commented on jax/interpreters/pxla.py:
```diff
@@ -339,11 +339,15 @@ def apply_parallel_primitive(prim, *args, **params):
-    shape = tuple(logical_size(dynamic_axis_env[name]) for name in axis_name)
+  axis_index_groups = params.pop('axis_index_groups')
+  if axis_index_groups is not None:
+    shape = (len(axis_index_groups[0]),)
```
I have another comment asking if groups are required to be the same size, I think this means they are

skye commented on jax/interpreters/pxla.py:
```diff
@@ -339,11 +339,15 @@ def apply_parallel_primitive(prim, *args, **params):
   else:
-    shape = (logical_size(dynamic_axis_env[axis_name]),)
+    logical_size = lambda frame: frame.hard_size * (frame.soft_size or 1)
+    if isinstance(axis_name, (list, tuple)):
```
I forgot you can specify multiple axis names in a collective. The current axis_index_groups API doesn't support that, so at least comment and ideally raise an error.

jekbradbury replied:
Yeah, good point. Strictly speaking it does work (indices in axis_index_groups will refer to the whole multiplexed axis formed from the multiple axis names) but I agree that we probably shouldn't do that and should just error instead.

skye replied:
Oh lol wow. I mean, it's arguably the jax way to just leave this undocumented behavior as-is, in case anyone needs it... I think I would leave it but your call.

jekbradbury replied:
I ended up leaving it.

mattjj replied:
+1 to leaving it, though I'll try to check in the code to make sure I understand the semantics.

skye reviewed (changes requested):
noice

jekbradbury commented on jax/lax/lax_parallel.py:
```diff
@@ -38,7 +38,7 @@
 ### parallel traceables
 
-def psum(x, axis_name):
+def psum(x, axis_name, axis_index_groups=None):
```
I also kind of like just `groups`, though verbosity seems less bad for nonobvious/"power user" features than it would be for something more central to the API. Matt suggested we start with an underscored name; WDYT about that option?
(I think my very weak overall preference is still `axis_index_groups`.)

skye replied:
Starting with an underscored name sounds fine to me, although this API seems sane enough to me that I don't think we need to be so defensive. My slight preference is no underscore, to avoid having an underscored param forever that everyone eventually uses anyway, but either is good with me.

> verbosity seems less bad for nonobvious/"power user" features
I agree, I just don't think the extra verbosity helps describe what this is. Or maybe it does? I can no longer see it with fresh eyes :) `axis_index_groups` is fine, it at least gives us a very specific name to call these things.

mattjj replied:
Yeah I agree this is looking solid enough that we might not need to be defensive.

On the name, I like James's idea!

jekbradbury commented on jax/lax/lax_parallel.py:
```diff
@@ -48,6 +48,9 @@ def psum(x, axis_name):
       ``pmap`` docstring for more details).
+    axis_index_groups: optional list of lists containing axis indices (e.g. for
+      an axis of size 4, [[0, 1], [2, 3]] would perform psums over the first
+      two and last two replicas).
```
Yes, they do. I think that's checked somewhere in XLA but it seems reasonable to also check it in JAX; where do you think the best place would be?

skye replied:
You probably have a better idea where to check than I do, but I think right in the main entrypoint (i.e. here) is good, so the primitives are always well-formed. You can make a little helper function `_validate_axis_index_groups` or whatever that you call in each collective function.

Also add it to the comment!

jekbradbury committed:
test other nesting order + imperfect nesting

jekbradbury committed:
update jaxpr.rst

jekbradbury committed:
handle None case

jekbradbury commented:
Addressed Skye's comments. @mattjj, want to take a look?

mattjj commented:
This is great! Thanks for pushing on this. And sorry for being slow to respond more. I'm still digging in but figure I should stream some thoughts to you as they come.

Here are some thoughts on the motivation in the PR message:

> Even in the most composable and multihost-friendly nested-pmap world that we can come up with, I think we'll still want to be able to express collectives that operate over subsets of the pmap axis.

I agree!

> In particular I think it's more natural, for example, for only the batchnorm layer to have to care about its distributed group size (whether to reduce over individual devices, neighboring pairs of devices, whole hosts, etc.), and keep the main pmap over the whole device axis, than for the entire model/training loop to need to know that (and use a pair of pmaps with separate axes).

I am less certain about this argument, but luckily I think you've also told me a stronger motivation. That is, my reading of the quoted text is that it's motivating this change as providing another way to express what we can already express with nested pmaps, just letting us have one axis (e.g. named "batch") rather than having two (e.g. "cross-host-batch" and "local-batch") while still being able to specify intra-host collectives. But I think the stronger version is to point to things we can't express even with nested pmaps (without requiring a sequence of separate pmap calls and some transposition). For example, with this proposal in a single pmap one can write `axis_index_groups=[[0, 2], [1, 3]]` and in another operation write `axis_index_groups=[[0, 1], [2, 3]]`, right? That seems like a sharper argument to me!

mattjj commented:
Btw, I found [this comment of yours](https://github.com/google/jax/pull/2382#issuecomment-622338934) extremely compelling, thanks so much for writing it out. I learned things!

mattjj commented on jax/interpreters/xla.py:
```diff
@@ -347,7 +347,13 @@ def write(v, node):
     elif eqn.primitive in parallel_translations:
       replica_groups = axis_groups(axis_env, eqn.params['axis_name'])
-      new_params = {k: v for k, v in eqn.params.items() if k != 'axis_name'}
+      axis_index_groups = eqn.params.get('axis_index_groups', None)
```
We shouldn't mutate `eqn.params`, and also it's a (related) smell to need a default here. We should make sure `axis_index_groups`, is _always_ populated from the start.

To be concrete: change this to `params['axis_index_groups']`?

jekbradbury replied:
I don't think `d.get(key, default)` mutates `d` (just checked).
Also, the reason `axis_index_groups` isn't always populated is because it's not present on all parallel primitives, just the ones that lower to `AllReduce`.  I believe it doesn't really make sense for `ppermute`, and while XLA `AllToAll` has a `replica_groups` attribute I'm not quite sure how we should expose that in the relevant lax wrappers.

mattjj replied:
Total brain fart on my part. I saw another `params.pop` elsewhere (where it was safe) and confused it with this one. Sorry for the noise!

Thanks for explaining the fact that it's not on all parallel primitives. That makes sense. I withdraw my suggestions!

mattjj commented on jax/lax/lax_parallel.py:
```diff
@@ -48,6 +48,9 @@ def psum(x, axis_name):
       ``pmap`` docstring for more details).
+    axis_index_groups: optional list of lists containing axis indices (e.g. for
+      an axis of size 4, [[0, 1], [2, 3]] would perform psums over the first
+      two and last two replicas). All groups must be the same size.
```
Maybe good to add that `axis_index_groups` must represent a partition of the integers {0, 1, ..., axis_size - 1} (i.e. it's a grouping of those integers into non-empty lists where each integer is included in a list exactly once) where the order of integers within each list matters.

jekbradbury replied:
Added both a note and a check.

mattjj commented on jax/lax/lax_parallel.py:
```diff
@@ -121,12 +134,24 @@ def pmin(x, axis_name):
+    return
+  len_0 = len(axis_index_groups[0])
+  if any(len(g) != len_0 for g in axis_index_groups):
+    raise ValueError("axis_index_groups must all be the same size")
```
Could check here that every index is used exactly once as well.

mattjj commented on jax/lax/lax_parallel.py:
```diff
@@ -257,8 +282,10 @@ def standard_pmap_primitive(name, multiple_results=False):
+def _allreduce_split_axis_rule(prim, reducer, vals, which_mapped, axis_name,
+                               axis_index_groups):
   assert tuple(which_mapped) == (True,)
+  assert axis_index_groups is None
```
Is this an assertion or rather an error? An assertion is best for representing "if this code wrong it's a jax implementation bug", but if this is an issue of validating the inputs in this case perhaps it's better to raise a ValueError or NotImplementedError. (I always think of this in terms of how `python -O` disables assertions.)

mattjj reviewed (approved):
Really great stuff! I have ~one key comment about mutating jaxpr eqns in-place, and~ (EDIT: that was just a mistake on my part) a couple questions / suggestions, but overall LGTM!

jekbradbury committed:
add note+check that groups  cover the index space

jekbradbury committed:
switch split_axis assert to NotImplementedError

jekbradbury committed:
update CHANGELOG

skye commented:
:tada:
